### ISSUE TITLE
Support batch secrets retrieval in Parameters module 

### ISSUE THREAD

**Issue by jeromevdl** [aws-powertools/powertools-lambda-java#1530]
**Is your feature request related to a problem? Please describe.**
N/A

**Describe the solution you'd like**
Add support for https://aws.amazon.com/about-aws/whats-new/2023/11/aws-secrets-manager-batch-retrieval-secrets/ in Parameters module.

**Comment by scottgerring** [comment]
This can be done in the v2 branch once #1403 is merged. It will be a good first issue at that point.

**Comment by jreijn** [comment]
@scottgerring @jeromevdl I would like to pick this up if possible. It seems #1403 has been merged, so good to go?

**Comment by jreijn** [comment]
I took a look at the code base and started with some minor improvements. I noticed that the getValue and the existing getMultipleValues exists, however the later on is now based on path. I'm thinking of renaming the second one to getMultipleValuesByPath and introduce a new method on the BaseProvider called getMultipleValuesByKey(List<String> keys). That way we can support both. I'll see if I can implement a solution in all the providers.

**Comment by scottgerring** [comment]
Hey @jreijn happy for you to pick this up - I will assign it to you. 
I'll take some time tomorrow morning to look at the current `...multipleValues` interface and write back!

**Comment by scottgerring** [comment]
## v2  `getMultipleValues` impls

### 1. `AppConfigProvider` - unsupported
https://github.com/aws-powertools/powertools-lambda-java/blob/3963d6d1f58c9618959f7a7044189a4a32d9b3fc/powertools-parameters/powertools-parameters-appconfig/src/main/java/software/amazon/lambda/powertools/parameters/appconfig/AppConfigProvider.java#L113-L117

### 2. `DynamoDbProvider`

Retrieves all values that share the same partition key. 

https://github.com/aws-powertools/powertools-lambda-java/blob/3963d6d1f58c9618959f7a7044189a4a32d9b3fc/powertools-parameters/powertools-parameters-dynamodb/src/main/java/software/amazon/lambda/powertools/parameters/dynamodb/DynamoDbProvider.java#L93-L99

### 3. `SSMProvider` 

Retrieves either everything at the given level of the `/parameter/hierarchy`, or everything at the given level plus `/parameter/hierarchy/nested/levels` when using `withRecursive`. 

https://github.com/aws-powertools/powertools-lambda-java/blob/3963d6d1f58c9618959f7a7044189a4a32d9b3fc/powertools-parameters/powertools-parameters-ssm/src/main/java/software/amazon/lambda/powertools/parameters/ssm/SSMProvider.java#L160-L166


Both the existing implementations - DDB and SSM provider - already vary a little - SSM is path-based, DDB is more like, common key. I'm a bit hesitant to change the existing method name because 1/ its not path-based in both cases and 2/ it's a breaking change against v1 which wouldn't add much. What do you think?

Maybe rather than changing the existing interface, we could add `getMultipleValuesByKey(keys)` on the `BaseProvider`,  provide a default impl that simply maps `getValue` over the keys, and then override in `SSMProvider` ? 

Thoughts @jreijn / @jeromevdl ?

**Comment by Attyuttam** [comment]
@phipag / @scottgerring  is this still open ?

**Comment by phipag** [comment]
Hey @Attyuttam, this is still open for implementation. Let me know if you have any questions and how you want to prioritize work on this. I see you are also assigned for SnapStart issues.

**Comment by Attyuttam** [comment]
Hi @phipag, have been on vacation, will start on these 2 tomorrow.

**Comment by Attyuttam** [comment]
Hi @phipag , i went through the conversation in this issue and based on my understanding, this is the approach I am planning to follow: 

1. create a method _getMultiple(List<String> keys)_ in ParamProvider.java
2. _getMultiple(List<String> keys)_ will have the implementation in BaseProvider.java in which I will loop over the _get(final String key)_ method. 

This way, all the keys will be retrieved from the cache if present.

Doubts:
1. Are you looking to cache the entire list of keys being requested ?
   - If that's the case, then I will need to change the approach as looping over get will not be a good idea then

**Comment by phipag** [comment]
Hey @Attyuttam,

let me clarify the scope of this issue.  We already have an interface in `BaseProvider` for a `getMultiple()` method. The scope of this issue about the Secrets module. @scottgerring's comment clarifies this a bit more: https://github.com/aws-powertools/powertools-lambda-java/issues/1530#issuecomment-1933504391. 

We should implement this method consistently with the other providers: https://github.com/aws-powertools/powertools-lambda-java/blob/4f82cd6ecab43a69a4e9b71601e49ce6cbf70148/powertools-parameters/powertools-parameters-secrets/src/main/java/software/amazon/lambda/powertools/parameters/secrets/SecretsProvider.java#L105-L107

The goal is to implement this method using the [BatchSecretGetValue](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_BatchGetSecretValue.html) API method introduced as per [this announcement.](https://aws.amazon.com/about-aws/whats-new/2023/11/aws-secrets-manager-batch-retrieval-secrets/). 

Let me know if this clarifies your implementation plan?

**Comment by phipag** [comment]
Here is an example in Powertools python docs: https://docs.aws.amazon.com/powertools/python/latest/utilities/parameters/#batch_secrets_providerpy

**Comment by Attyuttam** [comment]
@phipag this makes sense. Although, I was of the opinion that we will be using keys to get multiple values.

I was going through the BatchGetSecretValue API, it seems that i would need to create the Filter object in the BatchGetSecretValueRequest for which I am assuming the Key will be "name" and the Values will be the path passed ? Can you please confirm this.

**Comment by phipag** [comment]
> Maybe rather than changing the existing interface, we could add getMultipleValuesByKey(keys) on the BaseProvider, provide a default impl that simply maps getValue over the keys, and then override in SSMProvider ?

> I also noticed that lambda powertools in other languages do something similar. I'll make some small commits while I go along, so I can get some feedback along the way.

Yes, this is right with the exception that we need to accept a list of "paths". If you take a look at the Python implementation here the pattern becomes more clear: 

https://github.com/aws-powertools/powertools-lambda-python/blob/2c21763bf0e76e02728171d6ec5e6f068e1be3a9/aws_lambda_powertools/utilities/parameters/secrets.py#L156-L201

The question is if we need to change our base interfaces for this `BaseProvider` , `ParamProvider` which is the discussion we had above. I would say, let's not change it for now and instead only implement a `getMultiple(List names)` method in the SecretsProvider.

**Comment by Attyuttam** [comment]
Hi @phipag , sorry for the late reply, had been caught up with some work. 

In the SecretsProvider if we implement a method getMultiple(List<String> names) then we would also have to add the abstract methods in the BaseProvider and ParamProvider as there is no method which takes a list as of now. 

And we can then override in the SecretsProvider, we would have to write the implementation using BatchSecretsGetValue API

**Comment by phipag** [comment]
No problem @Attyuttam, happy to see you back! 

> we would also have to add the abstract methods in the BaseProvider and ParamProvider as there is no method which takes a list as of now.

I think this is not necessary. We can add the method directly to SecretsProvider without adding it to the base interface. And we update the Exception message in `getMultipleValues(String path)` to refer to the List overload instead. 

We can treat this as a special case in the SecretsProvider and mention this in the documentation. For a new major version release we can think about making a breaking change to the base interface since it is currently not flexible enough to support the different arguments for the different providers in a consistent way.

**Comment by Attyuttam** [comment]
@phipag thanks a lot for the explanation ! I think I get what the intention clearly is right now, I will update the Pr tomorrow
